### PR TITLE
Server version validation regex

### PR DESCRIPTION
**Pull request by respringer**
Provide validation based on regular expressions and fallback templates for docker image name creation.

### CODE REVIEW

File: charts/cass-operator-chart/templates/customresourcedefinition.yaml [hunk 1]
```diff
@@ -6039,15 +6039,7 @@ spec:
...
-              enum:
-              - 6.8.0
-              - 6.8.1
-              - 6.8.2
-              - 6.8.3
-              - 6.8.4
-              - 3.11.6
-              - 3.11.7
-              - 4.0.0
+              pattern: (6\.8\.\d)|(3\.11\.\d)|(4\.0\.\d)
```
Comment (sandoichi) [inline charts/cass-operator-chart/templates/customresourcedefinition.yaml:6042]: :100:

File: operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ type CassandraDatacenterSpec struct {
 
 	// Version string for config builder,
 	// used to generate Cassandra server configuration
-	// +kubebuilder:validation:Enum="6.8.0";"6.8.1";"6.8.2";"6.8.3";"6.8.4";"3.11.6";"3.11.7";"4.0.0"
+	// +kubebuilder:validation:Pattern=(6\.8\.\d)|(3\.11\.\d)|(4\.0\.\d)
```
Comment (sandoichi) [inline operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go:70]: :100:

File: operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types_test.go [hunk 3]
```diff
@@ -54,14 +56,24 @@ func Test_makeImage(t *testing.T) {
...
-			name: "test unknown version",
+			name: "test fallback",
 			args: args{
 				serverImage:   "",
 				serverType:    "dse",
-				serverVersion: "6.7.0",
+				serverVersion: "6.8.1234",
 			},
-			want:      "",
-			errString: "server 'dse' and version '6.7.0' do not work together",
```
Comment (jimdickinson) [inline operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types_test.go:64]: maybe should still keep this test case that should fail?

File: operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go [hunk 4]
```diff
@@ -67,7 +67,7 @@ type CassandraDatacenterSpec struct {
 
 	// Version string for config builder,
 	// used to generate Cassandra server configuration
-	// +kubebuilder:validation:Enum="6.8.0";"6.8.1";"6.8.2";"6.8.3";"6.8.4";"3.11.6";"3.11.7";"4.0.0"
+	// +kubebuilder:validation:Pattern=(6\.8\.\d+)|(3\.11\.\d+)|(4\.0\.\d+)
```
Comment (respringer) [inline operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go:70]: I was missing a plus sign. :facepalm:

File: operator/pkg/images/images.go [hunk 5]
```diff
@@ -199,7 +216,27 @@ func GetCassandraImage(serverType, version string) (string, error) {
...
+			if !IsDseVersionSupported(version) {
+				return "", fmt.Errorf("server 'dse' and version '%s' do not work together", version)
+			}
+			fallbackImageName = fmt.Sprintf("datastax/dse-server:%s", version)
+		} else {
+			if !IsOssVersionSupported(version) {
+				return "", fmt.Errorf("server 'cassandra' and version '%s' do not work together", version)
+			}
+			versionWithUnderscores := strings.Replace(version, ".", "_", -1)
+			fallbackImageName = fmt.Sprintf("datastax/cassandra-mgmtapi-%s:v0.0.1", versionWithUnderscores)
```
Comment (jimdickinson) [inline operator/pkg/images/images.go:232]: do we fix the 0.0.1 somewhere else?

Comment (respringer) [reply operator/pkg/images/images.go]: This has now been removed because of the newer naming scheme.

**Review by jimdickinson — APPROVED** [review]

**Pull request merged**